### ISSUE TITLE
Converted model from onnx to tensorflow has the same output for different input

### ISSUE THREAD

**Issue by ehsab** [onnx/onnx-tensorflow#281]
I am trying to convert pretrained resnet18 from torchvision model zoo to tensorflow using onnx.
I could convert the model, but the tf model always predicts the same thing for any input. 

I used the following code :

```
input_shape = (3, 224, 224)


def pytorch2onnx(src, des):
    import torch
    import torch.onnx as torch_onnx
    from torch.autograd import Variable
    from torchvision import models

    # Use this an input trace to serialize the model
    model_onnx_path = des
    # model = torch.load(src, map_location='cpu')
    model = models.resnet18(pretrained=True)

    model.train(False)

    # Export the model to an ONNX file
    dummy_input = Variable(torch.randn(512, *input_shape))
    output = torch_onnx.export(model,
                               dummy_input,
                               model_onnx_path,
                               verbose=False)


    print "Export of torch_model.onnx complete!"


def onnx2tf2(src, des):
    import onnx
    from onnx_tf.backend import prepare
    model = onnx.load(src)
    onnx.checker.check_model(model)
    print onnx.helper.printable_graph(model.graph)
    tf_rep = prepare(model, strict=False)
    tf_rep.export_graph(des)
```

Any idea?

**Comment by tjingrant** [comment]
Hi @ehsab , there's an issue template, if you fill in the required information, you'd get a useful response much faster. Specifically, would be very nice to have the onnx model in question.

**Comment by ehsab** [comment]
Thanks for letting me know. @tjingrant . I couldn't find the template but I attached the onnx model here.

[onnx model](https://1drv.ms/u/s!Asvf9EO9vyDxlJUI5eqRixWGSNCITw)

**Comment by fumihwh** [comment]
@ehsab How did you do inference? Any code will help.

**Comment by ehsab** [comment]
> @ehsab How did you do inference? Any code will help.

I use the following script. I should mention that this script works if I get the accuracy of original TF resnet50. It also works when I manually convert resnet18 to keras to TF. But, when I use onnx to convert pytorch original resnet18 to TF it doesn't work. It keeps predicting almost the same output for any input images.

p.s.
I already mentioned the code (see above) that I used to convert pytorch to TF with onnx.

```

import os
import time

import numpy as np
import tensorflow as tf
from tensorflow.python.platform import gfile

RESIZE_SIZE = 256
IMAGE_HEIGHT = 224
IMAGE_WIDTH = 224
BATCH_SIZE = 512
CPU = False

image_node_name = 'import/0:0'
output_node_name = 'import/add_8:0'
pb_file = '/tmp/resnet18.pb'
testing_dataset_dir = '/datasets/imagenet/imagenet_val/'


def get_filenames_and_classes(image_dir_root):
    class_names = []
    directories = []
    filenames = []
    corresponding_classes = []
    for filename in os.listdir(image_dir_root):
        path = os.path.join(image_dir_root, filename)
        if os.path.isdir(path):
            directories.append(path)
            class_names.append(filename)
    for i, directory in enumerate(directories):
        for filename in os.listdir(directory):
            filenames.append(os.path.join(directory, filename))
            corresponding_classes.append(class_names[i])
    class_names = map(str, sorted(map(str, class_names)))
    return filenames, corresponding_classes, class_names


def resize_image(input_image):
    # Resize image so that the shorter side is 256
    height_orig = tf.shape(input_image)[0]
    width_orig = tf.shape(input_image)[1]
    ratio_flag = tf.greater(height_orig, width_orig)  # True if height > width
    height = tf.where(ratio_flag, tf.cast(RESIZE_SIZE * height_orig / width_orig, tf.int32), RESIZE_SIZE)
    width = tf.where(ratio_flag, RESIZE_SIZE, tf.cast(RESIZE_SIZE * width_orig / height_orig, tf.int32))
    image = tf.image.resize_images(input_image, [height, width])
    return image


def parse_function(filename, label):
    image_string = tf.read_file(filename)
    image = tf.image.decode_jpeg(image_string, channels=3)

    # Resize_image
    image = resize_image(image)

    # Crop(random/center)
    height = IMAGE_HEIGHT
    width = IMAGE_WIDTH
    image_shape = tf.shape(image)
    h_offset = tf.cast((image_shape[0] - height) / 2, tf.int32)
    w_offset = tf.cast((image_shape[1] - width) / 2, tf.int32)
    image = tf.slice(image, [h_offset, w_offset, 0], [height, width, 3])

    # Normalization
    imagenet_mean = np.array([0.485, 0.456, 0.406], dtype=np.float32) * 255
    imagenet_std = np.array([0.229, 0.224, 0.225], dtype=np.float32) * 255
    image = (image - imagenet_mean) / imagenet_std

    return image, label


def get_dataset(image_dir_root, batch_size, num_parallel_calls=4, parsing_func=parse_function):
    filenames, corresponding_classes, class_names = get_filenames_and_classes(image_dir_root)
    dataset = tf.data.Dataset.from_tensor_slices((filenames, corresponding_classes))
    dataset = dataset.shuffle(buffer_size=len(filenames))
    dataset = dataset.map(parsing_func, num_parallel_calls=num_parallel_calls)
    dataset = dataset.batch(batch_size)
    iterator = dataset.make_initializable_iterator()
    next_element = iterator.get_next()
    init_op = iterator.initializer
    return init_op, next_element, class_names


if __name__ == '__main__':
    init_op, next_element, class_names = get_dataset(testing_dataset_dir, batch_size=BATCH_SIZE)
    class_dict = {}
    for i, class_name in enumerate(class_names):
        class_dict[class_name] = i
    if CPU:
        config = tf.ConfigProto(
            device_count={'GPU': 0,
                          'CPU': 1}
        )
    else:
        config = tf.ConfigProto(
            device_count={'GPU': 1,
                          'CPU': 1}
        )

    with tf.Session(config=config) as sess:
        with gfile.FastGFile(pb_file, 'rb') as f:
            graph_def = tf.GraphDef()
            graph_def.ParseFromString(f.read())
            g_in = tf.import_graph_def(graph_def)

        sess.run(init_op)
        total_num_of_images = 0
        num_of_correct_images = 0
        inference_times = []
        print("Starting Inference")
        while True:
            try:
                batch_images, batch_labels = sess.run(next_element)
                original_num_of_images = batch_images.shape[0]
                batch_labels = np.array([class_dict[batch_label] for batch_label in batch_labels])

                if original_num_of_images < BATCH_SIZE:
                    size = ((BATCH_SIZE - original_num_of_images),) + batch_images.shape[1:]
                    rest_images = np.random.randint(0, 255, size=size)
                    batch_images = np.concatenate((batch_images, rest_images))

                feed_dict = {image_node_name: np.reshape(batch_images, (512, 3, 224, 224))}
                fetch_list = [output_node_name]
                start_time = time.time()
                output = sess.run(fetch_list, feed_dict, options=tf.RunOptions(
                    trace_level=tf.RunOptions.FULL_TRACE))
                duration = (time.time() - start_time) * 1000.
                inference_times.append(duration)
                output[0] = output[0][:original_num_of_images]
                num_of_correct_images += np.sum(np.argmax(output[0], axis=-1) == batch_labels)
                total_num_of_images += len(output[0])
            except tf.errors.OutOfRangeError:
                break
        inference_times = np.array(inference_times)
        print("The average inference time per batch is ({} milliseconds) and per image is ({} milliseconds)".format(
            np.mean(inference_times),
            np.mean(inference_times) /
            float(BATCH_SIZE)))
        print("Measured Model Accuracy is {}".format(100. * (num_of_correct_images / float(total_num_of_images))))

```
Here the links to download [onxx](https://1drv.ms/u/s!Asvf9EO9vyDxlJUKI0Nc0Npa1hcbaw) model and [tf ](https://1drv.ms/u/s!Asvf9EO9vyDxlJULpX7XBYFQ_SFYtw)model. For pytorch model I used:  model = models.resnet18(pretrained=True) 

Any idea?

**Comment by ehsab** [comment]
> > @ehsab How did you do inference? Any code will help.
> 
> I use the following script. I should mention that this script works if I get the accuracy of original TF resnet50. It also works when I manually convert resnet18 to keras to TF. But, when I use onnx to convert pytorch original resnet18 to TF it doesn't work. It keeps predicting almost the same output for any input images.
> 
> p.s.
> I already mentioned the code (see above) that I used to convert pytorch to TF with onnx.
> 
> ```
> 
> import os
> import time
> 
> import numpy as np
> import tensorflow as tf
> from tensorflow.python.platform import gfile
> 
> RESIZE_SIZE = 256
> IMAGE_HEIGHT = 224
> IMAGE_WIDTH = 224
> BATCH_SIZE = 512
> CPU = False
> 
> image_node_name = 'import/0:0'
> output_node_name = 'import/add_8:0'
> pb_file = '/tmp/resnet18.pb'
> testing_dataset_dir = '/datasets/imagenet/imagenet_val/'
> 
> 
> def get_filenames_and_classes(image_dir_root):
>     class_names = []
>     directories = []
>     filenames = []
>     corresponding_classes = []
>     for filename in os.listdir(image_dir_root):
>         path = os.path.join(image_dir_root, filename)
>         if os.path.isdir(path):
>             directories.append(path)
>             class_names.append(filename)
>     for i, directory in enumerate(directories):
>         for filename in os.listdir(directory):
>             filenames.append(os.path.join(directory, filename))
>             corresponding_classes.append(class_names[i])
>     class_names = map(str, sorted(map(str, class_names)))
>     return filenames, corresponding_classes, class_names
> 
> 
> def resize_image(input_image):
>     # Resize image so that the shorter side is 256
>     height_orig = tf.shape(input_image)[0]
>     width_orig = tf.shape(input_image)[1]
>     ratio_flag = tf.greater(height_orig, width_orig)  # True if height > width
>     height = tf.where(ratio_flag, tf.cast(RESIZE_SIZE * height_orig / width_orig, tf.int32), RESIZE_SIZE)
>     width = tf.where(ratio_flag, RESIZE_SIZE, tf.cast(RESIZE_SIZE * width_orig / height_orig, tf.int32))
>     image = tf.image.resize_images(input_image, [height, width])
>     return image
> 
> 
> def parse_function(filename, label):
>     image_string = tf.read_file(filename)
>     image = tf.image.decode_jpeg(image_string, channels=3)
> 
>     # Resize_image
>     image = resize_image(image)
> 
>     # Crop(random/center)
>     height = IMAGE_HEIGHT
>     width = IMAGE_WIDTH
>     image_shape = tf.shape(image)
>     h_offset = tf.cast((image_shape[0] - height) / 2, tf.int32)
>     w_offset = tf.cast((image_shape[1] - width) / 2, tf.int32)
>     image = tf.slice(image, [h_offset, w_offset, 0], [height, width, 3])
> 
>     # Normalization
>     imagenet_mean = np.array([0.485, 0.456, 0.406], dtype=np.float32) * 255
>     imagenet_std = np.array([0.229, 0.224, 0.225], dtype=np.float32) * 255
>     image = (image - imagenet_mean) / imagenet_std
> 
>     return image, label
> 
> 
> def get_dataset(image_dir_root, batch_size, num_parallel_calls=4, parsing_func=parse_function):
>     filenames, corresponding_classes, class_names = get_filenames_and_classes(image_dir_root)
>     dataset = tf.data.Dataset.from_tensor_slices((filenames, corresponding_classes))
>     dataset = dataset.shuffle(buffer_size=len(filenames))
>     dataset = dataset.map(parsing_func, num_parallel_calls=num_parallel_calls)
>     dataset = dataset.batch(batch_size)
>     iterator = dataset.make_initializable_iterator()
>     next_element = iterator.get_next()
>     init_op = iterator.initializer
>     return init_op, next_element, class_names
> 
> 
> if __name__ == '__main__':
>     init_op, next_element, class_names = get_dataset(testing_dataset_dir, batch_size=BATCH_SIZE)
>     class_dict = {}
>     for i, class_name in enumerate(class_names):
>         class_dict[class_name] = i
>     if CPU:
>         config = tf.ConfigProto(
>             device_count={'GPU': 0,
>                           'CPU': 1}
>         )
>     else:
>         config = tf.ConfigProto(
>             device_count={'GPU': 1,
>                           'CPU': 1}
>         )
> 
>     with tf.Session(config=config) as sess:
>         with gfile.FastGFile(pb_file, 'rb') as f:
>             graph_def = tf.GraphDef()
>             graph_def.ParseFromString(f.read())
>             g_in = tf.import_graph_def(graph_def)
> 
>         sess.run(init_op)
>         total_num_of_images = 0
>         num_of_correct_images = 0
>         inference_times = []
>         print("Starting Inference")
>         while True:
>             try:
>                 batch_images, batch_labels = sess.run(next_element)
>                 original_num_of_images = batch_images.shape[0]
>                 batch_labels = np.array([class_dict[batch_label] for batch_label in batch_labels])
> 
>                 if original_num_of_images < BATCH_SIZE:
>                     size = ((BATCH_SIZE - original_num_of_images),) + batch_images.shape[1:]
>                     rest_images = np.random.randint(0, 255, size=size)
>                     batch_images = np.concatenate((batch_images, rest_images))
> 
>                 feed_dict = {image_node_name: np.reshape(batch_images, (512, 3, 224, 224))}
>                 fetch_list = [output_node_name]
>                 start_time = time.time()
>                 output = sess.run(fetch_list, feed_dict, options=tf.RunOptions(
>                     trace_level=tf.RunOptions.FULL_TRACE))
>                 duration = (time.time() - start_time) * 1000.
>                 inference_times.append(duration)
>                 output[0] = output[0][:original_num_of_images]
>                 num_of_correct_images += np.sum(np.argmax(output[0], axis=-1) == batch_labels)
>                 total_num_of_images += len(output[0])
>             except tf.errors.OutOfRangeError:
>                 break
>         inference_times = np.array(inference_times)
>         print("The average inference time per batch is ({} milliseconds) and per image is ({} milliseconds)".format(
>             np.mean(inference_times),
>             np.mean(inference_times) /
>             float(BATCH_SIZE)))
>         print("Measured Model Accuracy is {}".format(100. * (num_of_correct_images / float(total_num_of_images))))
> ```
> Here the links to download [onxx](https://1drv.ms/u/s!Asvf9EO9vyDxlJUKI0Nc0Npa1hcbaw) model and [tf ](https://1drv.ms/u/s!Asvf9EO9vyDxlJULpX7XBYFQ_SFYtw)model. For pytorch model I used: model = models.resnet18(pretrained=True)
> 
> Any idea?

**Comment by BranyonApel** [comment]
This same issue still seems to exist almost exactly as described above. A model trained in pytorch will function but when used in tensorflow via onnx the model does not respond to any change input, just a fixed output. @ehsab did you ever find a solution to this?
